libnmstate/nm/route.py

```diff
@@ -0,0 +1,129 @@
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+#
+
+from operator import itemgetter
+from libnmstate.error import NmstateInternalError
+from libnmstate.schema import Route
+
+from libnmstate import iplib
+
+NM_MAIN_ROUTE_TABLE_ID = 0
+
+
+def get_running(acs_and_ip_cfgs):
+    """
+    Query running routes
+    The acs_and_ip_cfgs should be generate to generate a tuple:
+        NM.NM.ActiveConnection, NM.IPConfig
+    """
+    routes = []
+    for (active_connection, ip_cfg) in acs_and_ip_cfgs:
+        if not ip_cfg.props.routes:
+            continue
+        iface_name = _get_iface_name(active_connection)
+        if not iface_name:
+            raise NmstateInternalError(
+                'Got connection {} has not interface name'.format(
+                    active_connection.get_id()))
+        for nm_route in ip_cfg.props.routes:
+            table_id = _get_per_route_table_id(nm_route)
+            route_entry = _nm_route_to_route(
+                nm_route,
+                table_id,
+                iface_name)
+            if route_entry:
+                routes.append(route_entry)
+    routes.sort(key=itemgetter(Route.TABLE_ID,
+                               Route.NEXT_HOP_INTERFACE,
+                               Route.DESTINATION))
+    return routes
+
+
+def get_config(acs_and_ip_profiles):
+    """
+    Query running routes
+    The acs_and_ip_profiles should be generate to generate a tuple:
+        NM.NM.ActiveConnection, NM.SettingIPConfig
+    """
+    routes = []
+    for (active_connection, ip_profile) in acs_and_ip_profiles:
+        if not ip_profile.props.routes:
+            continue
+        iface_name = _get_iface_name(active_connection)
+        if not iface_name:
+            raise NmstateInternalError(
+                'Got connection {} has not interface name'.format(
+                    active_connection.get_id()))
+        table_id = _get_cfg_default_route_table_id(ip_profile)
+        # NM supports multiple route table in single profile:
+        #   https://bugzilla.redhat.com/show_bug.cgi?id=1436531
+        # The `ipv4.route-table` and `ipv6.route-table` will be the default
+        # table id for static routes and auto routes. But each static route can
+        # still specify route table id.
+        for nm_route in ip_profile.props.routes:
+            table = nm_route.get_attribute('table')
```

Comment:
You could change ` _get_per_route_table_id(nm_route)` to also take an optional default ID argument and then re-use it here, too.

Reply:
Sure. Will do.